### ISSUE TITLE
synchronous sources

### ISSUE THREAD

**Issue by nichoth** [pull-stream/pull-stream-faq#3]
Why are both synchronous and asynchronous sources supported? For performance? It mixes two very different patterns, and makes interoperability more difficult.

**Comment by dominictarr** [comment]
Yes, performance. For a sink to read inside a loop where possible is much lighter than using `nextTick` (etc)
if throughs also always used nextTick, then that would add up on a long pipeline. If the source is async.

However, since it always uses callbacks even when sync, it's not really that much different. Also, supporting sync actually has the benefit that now you can reason accurately about order, which makes testing different orders easier.

**Comment by dominictarr** [comment]
anyway, if you want interop, you can always convert a sync pull-stream into an async one and the api still works, so if you want it that way, you could make a say, `pull-async` module that checked if the read callback had been called sync, and then applying `nextTick`. Then you'd have your async stream.
